src/basic/cgroup-util.c

```diff
@@ -152,6 +152,17 @@ int cg_read_pidref(FILE *f, PidRef *ret, CGroupFlags flags) {
                 r = pidref_set_pid(ret, pid);
                 if (r >= 0)
                         return 1;
+                /* Opening pidfds for non thread group leaders only works from 6.9 onwards. On older kernels
+                 * pidfd_open() fails with EINVAL. Since we might read non thread group leader IDs from
+                 * cgroup.threads and try to open pidfd's for them, let's ignore EINVAL from pidref_set_pid()
+                 * and fallback to using the pid instead. */
+                if (r == -EINVAL) {
```

Comment:
I don't think we should do this generically in cg_read_pidref. PID 1's process tracking logic doesn't take threads into account, so this feels intrusive.

Instead, cg_kill_items should take a do_pidfd flag?

Reply:
I think checking here is OK. But how about `if (r == -EINVAL && !is_main_thread())` ?

Reply:
It is completely pointless for cg_kill_items, and for other callers it means unexpected result. Hence why bother?